File: backend/routes/businesses.py
"""Business listing CRUD and photo proxy routes.

Provides endpoints for listing, searching, creating, updating, and deleting
businesses, plus a photo proxy that serves Google Place images with
category-based SVG fallbacks.
"""
import io
import logging
from typing import List, Optional
from datetime import datetime
import re
from fastapi import APIRouter, HTTPException, status, Depends, Query
from fastapi.responses import StreamingResponse
from bson import ObjectId

from backend.models.business import (
    Business,
    BusinessCreate,
    BusinessProfileUpdate,
    BusinessUpdate,
    CategoryEnum,
)
from backend.models.user import User
from backend.models.auth import get_current_user
from backend.database.document_store import get_businesses_collection, DatabaseUnavailableError
from backend.services.business_metadata import (
    derive_known_for,
    generate_short_description,
    normalize_business_metadata,
    normalize_image_urls,
)
from backend.services.photo_proxy import build_category_placeholder_bytes, build_stream, get_photo_payload
from backend.routes.discovery import discover_businesses
from backend.utils.security import normalize_optional_url, normalize_text_list, sanitize_text

router = APIRouter()

logger = logging.getLogger(__name__)

BUSINESS_LIST_PROJECTION = {
    "name": 1,
    "category": 1,
    "description": 1,
    "address": 1,
    "city": 1,
    "location": 1,
    "rating_average": 1,
    "total_reviews": 1,
    "has_deals": 1,
    "owner_id": 1,
    "place_id": 1,
    "image": 1,
    "image_url": 1,
    "image_urls": 1,
    "primary_image_url": 1,
    "short_description": 1,
    "known_for": 1,
    "is_claimed": 1,
    "claim_status": 1,
    "credibility_score": 1,
    "live_visibility_score": 1,
    "local_confidence": 1,
    "is_active_today": 1,
    "checkins_today": 1,
    "trending_score": 1,
    "last_activity_at": 1,
    "created_at": 1,
    "phone": 1,
    "email": 1,
    "website": 1,
}

def _oid(raw_id: str) -> ObjectId:
    if not ObjectId.is_valid(raw_id):
        raise HTTPException(
            status_code=status.HTTP_400_BAD_REQUEST,
            detail="Invalid business ID format"
        )
    return ObjectId(raw_id)

async def _business_or_404(business_id: str) -> dict:
    businesses_collection = get_businesses_collection()
    business = await businesses_collection.find_one({"_id": _oid(business_id)})
    if not business:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail="Business not found"
        )
    return business

def business_helper(business) -> dict:
    if business:
        business.setdefault("image_url", business.pop("image", ""))
        business.setdefault(
            "primary_image_url",
            business.get("image_url") or ((business.get("image_urls") or [""])[0] if business.get("image_urls") else ""),
        )
        if not business.get("short_description") or not business.get("known_for") or not business.get("image_urls"):
            normalize_business_metadata(business)
        business["id"] = str(business.pop("_id"))
        if "rating_average" in business:
            business["rating"] = business.pop("rating_average")
        if "total_reviews" in business:
            business["review_count"] = business.pop("total_reviews")
        business.setdefault("rating", 0.0)
        business.setdefault("review_count", 0)
        business.setdefault("has_deals", False)
        if "owner_id" in business:
            business["owner_id"] = str(business["owner_id"])
    return business

def _normalize_business_url(value: Optional[str], label: str) -> Optional[str]:
    try:
        return normalize_optional_url(value)
    except ValueError as exc:
        raise HTTPException(
            status_code=status.HTTP_400_BAD_REQUEST,
            detail=f"Invalid {label}: {exc}",
        ) from exc

def _normalize_business_image_urls(values: Optional[list[str]], primary_image: Optional[str] = None) -> list[str]:
    cleaned: list[str] = []
    for value in values or []:
        normalized = _normalize_business_url(value, "image URL")
        if normalized:
            cleaned.append(normalized)
    return normalize_image_urls(cleaned, primary_image=primary_image or "")

def _sanitize_business_update(data: dict) -> dict:
    """Sanitize business-owner controlled fields before persistence."""
    text_limits = {
        "name": 200,
        "description": 1000,
        "address": 300,
        "city": 100,
        "phone": 50,
        "email": 254,
        "short_description": 160,
    }
    for key, limit in text_limits.items():
        if key in data and data[key] is not None:
            data[key] = sanitize_text(data[key], max_length=limit)
    if "website" in data:
        data["website"] = _normalize_business_url(data.get("website"), "website URL")
    if "image_url" in data:
        data["image_url"] = _normalize_business_url(data.get("image_url"), "image URL")
    if "image_urls" in data and data["image_urls"] is not None:
        data["image_urls"] = _normalize_business_image_urls(
            data.get("image_urls"),
            primary_image=data.get("image_url", ""),
        )
    if "known_for" in data and data["known_for"] is not None:
        data["known_for"] = normalize_text_list(data.get("known_for"), limit=6, max_item_length=24)
    return data

@router.get("/photos")
async def get_business_photo(
    place_id: str = Query(..., min_length=3, max_length=256),
    maxwidth: int = Query(800, ge=120, le=1600),
):
    """Proxy a business photo by Google Place ID (GET /api/photos).

    Resolves the image from Google's photo API, an OG image on the
    business website, or returns an SVG placeholder. Responses are
    cached in memory and on disk.

    Args:
        place_id (str): Google Places place ID.
        maxwidth (int): Desired image width in pixels (120-1600).

    Returns:
        StreamingResponse: The image payload with cache headers.
    """
    try:
        businesses_collection = get_businesses_collection()
        content_type, payload = await get_photo_payload(businesses_collection, place_id.strip(), maxwidth)
        stream, headers = build_stream(content_type, payload)
    except DatabaseUnavailableError:
        logger.exception("DB error serving photo for place_id=%s", place_id)
        content_type, payload = build_category_placeholder_bytes(label="V")
        stream = io.BytesIO(payload)
        headers = {"Cache-Control": "no-store", "Content-Length": str(len(payload))}
    return StreamingResponse(stream, media_type=content_type, headers=headers)

@router.get("/businesses", response_model=List[Business])
async def get_businesses(
    category: Optional[CategoryEnum] = None,
    city: Optional[str] = Query(None, max_length=100),
    search: Optional[str] = Query(None, max_length=120),
    owner_id: Optional[str] = Query(None, max_length=128),
    min_rating: Optional[float] = Query(None, ge=0, le=5),
    skip: int = Query(0, ge=0),
    limit: int = Query(50, ge=1, le=100)
):
    """List businesses with optional filtering (GET /api/businesses).

    Supports filtering by category, city (case-insensitive regex), minimum
    rating, and full-text search across name and description.

    Returns:
        List[Business]: Filtered and paginated business listings.
    """
    businesses_collection = get_businesses_collection()
    query = {}
    if category:
        query["category"] = category
    if owner_id:
        query["owner_id"] = owner_id
    if city:
        query["city"] = {"$regex": re.escape(city), "$options": "i"}
    if min_rating is not None:
        query["rating_average"] = {"$gte": min_rating}
    if search:
        escaped_search = re.escape(search)
        query["$or"] = [
            {"name": {"$regex": escaped_search, "$options": "i"}},
            {"description": {"$regex": escaped_search, "$options": "i"}}
        ]
    cursor = businesses_collection.find(query, BUSINESS_LIST_PROJECTION).skip(skip).limit(limit)
    businesses = await cursor.to_list(length=limit)
    return [business_helper(business) for business in businesses]

@router.get("/businesses/feed")
async def get_businesses_feed(
    page: int = Query(1, ge=1),
    page_size: int = Query(10, ge=1, le=50),
    category: Optional[CategoryEnum] = None,
    search: Optional[str] = Query(None, max_length=120),
    sort_by: Optional[str] = Query(None, description="Sort by: rating, reviews, newest"),
):
    """Paginated business feed with sorting (GET /api/businesses/feed).

    Returns a page of businesses along with total count and pagination metadata.

    Returns:
        dict: ``{"businesses": [...], "total": int, "page": int, "page_size": int, "has_more": bool}``
    """
    businesses_collection = get_businesses_collection()

    query = {}
    if category:
        query["category"] = category
    if search:
        escaped_search = re.escape(search)
        query["$or"] = [
            {"name": {"$regex": escaped_search, "$options": "i"}},
            {"description": {"$regex": escaped_search, "$options": "i"}}
        ]

    sort_field = "created_at"
    sort_dir = -1
    if sort_by == "rating":
        sort_field = "rating_average"
        sort_dir = -1
    elif sort_by == "reviews":
        sort_field = "total_reviews"
        sort_dir = -1
    elif sort_by == "newest":
        sort_field = "created_at"
        sort_dir = -1

    skip = (page - 1) * page_size
    cursor = businesses_collection.find(query, BUSINESS_LIST_PROJECTION).sort(sort_field, sort_dir).skip(skip).limit(page_size + 1)
    page_items = await cursor.to_list(length=page_size + 1)
    has_more = len(page_items) > page_size
    businesses = page_items[:page_size]

    return {
        "businesses": [business_helper(b) for b in businesses],
        "total": skip + len(businesses) + (1 if has_more else 0),
        "page": page,
        "page_size": page_size,
        "has_more": has_more,
    }

@router.get("/businesses/nearby", response_model=List[Business])
async def get_nearby_businesses(
    lat: float = Query(..., ge=-90, le=90, description="Latitude"),
    lng: float = Query(..., ge=-180, le=180, description="Longitude"),
    radius: float = Query(10, ge=1, le=100, description="Search radius in kilometers"),
    category: Optional[CategoryEnum] = None,
    min_rating: Optional[float] = Query(None, ge=0, le=5),
    limit: int = Query(50, ge=1, le=100)
):
    """Find businesses near a given location (GET /api/businesses/nearby).

    Delegates to the discovery engine and optionally filters by minimum
    rating after retrieval.

    Returns:
        List[Business]: Businesses sorted by canonical ranking within the radius.
    """
    businesses = await discover_businesses(
        lat=lat,
        lng=lng,
        radius=radius,
        category=category.value if category else None,
        limit=limit,
        sort_mode="canonical",
        refresh=False,
    )
    if min_rating is not None:
        businesses = [business for business in businesses if (business.get("rating") or 0) >= min_rating]
    return businesses[:limit]

@router.get("/businesses/{business_id}", response_model=Business)
async def get_business(business_id: str):
    """Retrieve a single business by ID (GET /api/businesses/{business_id}).

    Returns:
        Business: The requested business.

    Raises:
        HTTPException: 404 if the business does not exist.
    """
    return business_helper(await _business_or_404(business_id))

@router.post("/businesses", response_model=Business, status_code=status.HTTP_201_CREATED)
async def create_business(
    business_data: BusinessCreate,
    current_user: User = Depends(get_current_user)
):
    """Create a new business listing (POST /api/businesses).

    Only users with the ``business_owner`` role may create businesses.
    Auto-generates ``short_description`` and ``known_for`` tags from the
    category and address.

    Returns:
        Business: The newly created business.

    Raises:
        HTTPException: 403 if the user is not a business owner.
    """
    if current_user.role != "business_owner":
        raise HTTPException(
            status_code=status.HTTP_403_FORBIDDEN,
            detail="Only business accounts can create businesses"
        )
    businesses_collection = get_businesses_collection()
    business_dict = {
        "owner_id": current_user.id,
        "name": business_data.name,
        "category": business_data.category,
        "description": business_data.description,
        "address": business_data.address,
        "city": business_data.city,
        "location": business_data.location.model_dump(),
        "rating_average": 0.0,
        "total_reviews": 0,
        "phone": business_data.phone,
        "email": business_data.email,
        "website": business_data.website,
        "image_url": business_data.image_url,
        "image_urls": normalize_image_urls(business_data.image_urls, primary_image=business_data.image_url or ""),
        "short_description": generate_short_description(
            category=business_data.category,
            address=business_data.address,
            city=business_data.city,
            existing=business_data.short_description or business_data.description,
        ),
        "known_for": derive_known_for(
            category=business_data.category,
            existing=business_data.known_for,
        ),
        "created_at": datetime.utcnow()
    }
    _sanitize_business_update(business_dict)
    normalize_business_metadata(business_dict)
    result = await businesses_collection.insert_one(business_dict)
    created_business = await businesses_collection.find_one({"_id": result.inserted_id})
    return business_helper(created_business)

@router.put("/businesses/{business_id}", response_model=Business)
async def update_business(
    business_id: str,
    business_data: BusinessUpdate,
    current_user: User = Depends(get_current_user)
):
    """Update a business listing (PUT /api/businesses/{business_id}).

    Only the business owner may update their listing. Re-derives
    ``short_description`` and ``known_for`` when relevant fields change.

    Returns:
        Business: The updated business.

    Raises:
        HTTPException: 403 if the user is not the business owner.
        HTTPException: 404 if the business does not exist.
    """
    businesses_collection = get_businesses_collection()
    business_key = _oid(business_id)
    business = await businesses_collection.find_one({"_id": business_key})
    if not business:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail="Business not found")
    if str(business["owner_id"]) != current_user.id:
        raise HTTPException(
            status_code=status.HTTP_403_FORBIDDEN,
            detail="Not authorized to update this business"
        )
    update_data = {k: v for k, v in business_data.model_dump(exclude_unset=True).items() if v is not None}
    _sanitize_business_update(update_data)
    if not update_data:
        raise HTTPException(
            status_code=status.HTTP_400_BAD_REQUEST,
            detail="No valid fields to update"
        )
    if "location" in update_data and hasattr(update_data["location"], "model_dump"):
        update_data["location"] = update_data["location"].model_dump()
    if "image_urls" in update_data or "image_url" in update_data:
        existing_images = update_data.get("image_urls") if "image_urls" in update_data else business.get("image_urls", [])
        primary_image = update_data.get("image_url", business.get("image_url", ""))
        update_data["image_urls"] = normalize_image_urls(existing_images, primary_image=primary_image)
    if "known_for" in update_data:
        update_data["known_for"] = derive_known_for(
            category=update_data.get("category", business.get("category", "")),
            google_types=business.get("google_types", []),
            existing=update_data["known_for"],
        )
    if "short_description" in update_data or "description" in update_data:
        update_data["short_description"] = generate_short_description(
            category=update_data.get("category", business.get("category", "")),
            address=update_data.get("address", business.get("address", "")),
            city=update_data.get("city", business.get("city", "")),
            existing=update_data.get(
                "short_description",
                update_data.get("description", business.get("short_description") or business.get("description", "")),
            ),
        )
    normalized_business = {**business, **update_data}
    normalize_business_metadata(normalized_business)
    for key in ("description", "short_description", "known_for", "image_url", "image_urls", "primary_image_url"):
        if key in normalized_business:
            update_data[key] = normalized_business[key]
    await businesses_collection.update_one(
        {"_id": business_key},
        {"$set": update_data}
    )
    updated_business = await businesses_collection.find_one({"_id": business_key})
    return business_helper(updated_business)

@router.put("/businesses/{business_id}/profile", response_model=Business)
async def update_business_profile(
    business_id: str,
    profile_data: BusinessProfileUpdate,
    current_user: User = Depends(get_current_user)
):
    """Update owner-editable profile fields (PUT /api/businesses/{business_id}/profile).

    Restricted to the claimed business owner. Only ``short_description``
    and ``known_for`` tags may be updated through this endpoint.

    Returns:
        Business: The updated business.

    Raises:
        HTTPException: 403 if the user is not the claimed owner.
    """
    businesses_collection = get_businesses_collection()
    business_key = _oid(business_id)
    business = await businesses_collection.find_one({"_id": business_key})
    if not business:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail="Business not found")

    if not business.get("is_claimed") or str(business.get("owner_id") or "") != current_user.id:
        raise HTTPException(
            status_code=status.HTTP_403_FORBIDDEN,
            detail="Only the claimed owner can edit business profile details"
        )

    payload = profile_data.model_dump(exclude_unset=True)
    _sanitize_business_update(payload)
    update_data = {}

    if "short_description" in payload:
        update_data["short_description"] = generate_short_description(
            category=business.get("category", ""),
            address=business.get("address", ""),
            city=business.get("city", ""),
            existing=payload.get("short_description", ""),
        )

    if "known_for" in payload:
        update_data["known_for"] = derive_known_for(
            category=business.get("category", ""),
            google_types=business.get("google_types", []),
            existing=payload.get("known_for", []),
        )

    if not update_data:
        raise HTTPException(
            status_code=status.HTTP_400_BAD_REQUEST,
            detail="No valid profile fields to update"
        )

    update_data["updated_at"] = datetime.utcnow()
    normalized_business = {**business, **update_data}
    normalize_business_metadata(normalized_business)
    update_data["short_description"] = normalized_business.get("short_description")
    update_data["known_for"] = normalized_business.get("known_for", [])
    update_data["description"] = normalized_business.get("description", business.get("description", ""))

    await businesses_collection.update_one(
        {"_id": business_key},
        {"$set": update_data}
    )

    updated_business = await businesses_collection.find_one({"_id": business_key})
    return business_helper(updated_business)

@router.delete("/businesses/{business_id}", status_code=status.HTTP_204_NO_CONTENT)
async def delete_business(
    business_id: str,
    current_user: User = Depends(get_current_user)
):
    """Delete a business listing (DELETE /api/businesses/{business_id}).

    Only the business owner may delete their listing.

    Raises:
        HTTPException: 403 if the user is not the business owner.
        HTTPException: 404 if the business does not exist.
    """
    businesses_collection = get_businesses_collection()
    business_key = _oid(business_id)
    business = await businesses_collection.find_one({"_id": business_key})
    if not business:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail="Business not found")
    if str(business["owner_id"]) != current_user.id:
        raise HTTPException(
            status_code=status.HTTP_403_FORBIDDEN,
            detail="Not authorized to delete this business"
        )
    await businesses_collection.delete_one({"_id": business_key})
    return None

File: backend/src/routes/businesses.rs
use crate::{
    errors::{AppError, Result},
    middleware::auth::AuthUser,
    models::business::{BusinessCreate, BusinessSearchQuery, BusinessUpdate},
    routes::support::{
        business_lat_lng, eq, ilike_or_filter, is_true, limit, normalize_business, offset, order,
        select_all, value_str, QueryParams,
    },
    security,
    state::AppState,
};
use axum::{
    body::Bytes,
    extract::{Path, Query, State},
    http::{header, HeaderValue, StatusCode},
    response::{IntoResponse, Response},
    routing::get,
    Json, Router,
};
use chrono::Utc;
use serde_json::{json, Map, Value};
use std::sync::Arc;
use validator::Validate;

pub fn router() -> Router<Arc<AppState>> {
    Router::new()
        .route("/businesses", get(list_businesses).post(create_business))
        .route("/businesses/nearby", get(nearby_businesses))
        .route(
            "/businesses/:id/profile",
            axum::routing::put(update_business_profile),
        )
        .route(
            "/businesses/:id",
            get(get_business)
                .put(update_business)
                .delete(delete_business),
        )
        .route("/businesses/:id/photo", get(get_business_photo))
        .route("/photos", get(get_google_place_photo))
}

async fn list_businesses(
    State(state): State<Arc<AppState>>,
    Query(params): Query<BusinessSearchQuery>,
) -> Result<impl IntoResponse> {
    let rows = query_businesses(&state, &params, params.limit.unwrap_or(100).clamp(1, 200)).await?;
    Ok(Json(rows))
}

async fn nearby_businesses(
    State(state): State<Arc<AppState>>,
    Query(mut params): Query<BusinessSearchQuery>,
) -> Result<impl IntoResponse> {
    params.sort = Some("distance".into());
    let rows = query_businesses(&state, &params, params.limit.unwrap_or(50).clamp(1, 100)).await?;
    Ok(Json(rows))
}

async fn get_business(
    State(state): State<Arc<AppState>>,
    Path(id): Path<String>,
) -> Result<impl IntoResponse> {
    let id = security::validate_uuid_id(&id, "business ID")?;
    let row = find_business(&state, &id).await?;
    Ok(Json(normalize_business(row)))
}

async fn create_business(
    State(state): State<Arc<AppState>>,
    auth_user: AuthUser,
    Json(payload): Json<BusinessCreate>,
) -> Result<impl IntoResponse> {
    payload
        .validate()
        .map_err(|e| AppError::BadRequest(e.to_string()))?;

    if let (Some(lat), Some(lng)) = (payload.lat, payload.lng) {
        security::validate_lat_lng(lat, lng)?;
    }
    let website = security::normalize_url(payload.website.as_deref(), 500, false)?;
    let now = Utc::now().to_rfc3339();

    let mut body = Map::new();
    body.insert(
        "name".into(),
        json!(security::sanitize_text(&payload.name, 200)),
    );
    body.insert(
        "address".into(),
        json!(security::sanitize_text(&payload.address, 300)),
    );
    body.insert("is_verified".into(), json!(false));
    body.insert("is_claimed".into(), json!(false));
    body.insert("owner_id".into(), json!(auth_user.id));
    body.insert("review_count".into(), json!(0));
    body.insert("photos".into(), json!([]));
    body.insert("known_for".into(), json!([]));
    body.insert("created_at".into(), json!(now));
    body.insert("updated_at".into(), json!(now));

    insert_optional_text(&mut body, "category", payload.category.as_deref(), 80);
    insert_optional_text(&mut body, "city", payload.city.as_deref(), 120);
    insert_optional_text(&mut body, "state", payload.state.as_deref(), 80);
    insert_optional_text(&mut body, "zip_code", payload.zip_code.as_deref(), 20);
    insert_optional_text(&mut body, "phone", payload.phone.as_deref(), 40);
    insert_optional_text(
        &mut body,
        "description",
        payload.description.as_deref(),
        1000,
    );
    if let Some(website) = website {
        body.insert("website".into(), json!(website));
    }
    if let (Some(lat), Some(lng)) = (payload.lat, payload.lng) {
        body.insert(
            "location".into(),
            json!({ "type": "Point", "coordinates": [lng, lat] }),
        );
    }

    let created = state
        .db
        .supabase
        .insert_json("businesses", Value::Object(body))
        .await?;

    Ok((StatusCode::CREATED, Json(normalize_business(created))))
}

async fn update_business(
    State(state): State<Arc<AppState>>,
    auth_user: AuthUser,
    Path(id): Path<String>,
    Json(payload): Json<BusinessUpdate>,
) -> Result<impl IntoResponse> {
    let id = security::validate_uuid_id(&id, "business ID")?;
    let existing = find_business(&state, &id).await?;
    ensure_business_owner(&existing, &auth_user)?;

    if let (Some(lat), Some(lng)) = (payload.lat, payload.lng) {
        security::validate_lat_lng(lat, lng)?;
    }
    let website = security::normalize_url(payload.website.as_deref(), 500, false)?;

    let mut body = Map::new();
    body.insert("updated_at".into(), json!(Utc::now().to_rfc3339()));
    insert_optional_text(&mut body, "name", payload.name.as_deref(), 200);
    insert_optional_text(&mut body, "category", payload.category.as_deref(), 80);
    insert_optional_text(&mut body, "address", payload.address.as_deref(), 300);
    insert_optional_text(&mut body, "city", payload.city.as_deref(), 120);
    insert_optional_text(&mut body, "state", payload.state.as_deref(), 80);
    insert_optional_text(&mut body, "zip_code", payload.zip_code.as_deref(), 20);
    insert_optional_text(&mut body, "phone", payload.phone.as_deref(), 40);
    insert_optional_text(
        &mut body,
        "description",
        payload.description.as_deref(),
        1000,
    );
    if let Some(website) = website {
        body.insert("website".into(), json!(website));
    }
    if let Some(hours) = payload.hours {
        body.insert(
            "hours".into(),
            serde_json::to_value(hours).unwrap_or(Value::Null),
        );
    }
    if let (Some(lat), Some(lng)) = (payload.lat, payload.lng) {
        body.insert(
            "location".into(),
            json!({ "type": "Point", "coordinates": [lng, lat] }),
        );
    }

    let updated = state
        .db
        .supabase
        .update_json("businesses", &[eq("id", &id)], Value::Object(body))
        .await?;

    let row = updated
        .into_iter()
        .next()
        .ok_or_else(|| AppError::NotFound("Business not found".into()))?;
    Ok(Json(normalize_business(row)))
}

async fn update_business_profile(
    State(state): State<Arc<AppState>>,
    auth_user: AuthUser,
    Path(id): Path<String>,
    Json(payload): Json<Value>,
) -> Result<impl IntoResponse> {
    let id = security::validate_uuid_id(&id, "business ID")?;
    let existing = find_business(&state, &id).await?;
    ensure_business_owner(&existing, &auth_user)?;

    let mut body = Map::new();
    body.insert("updated_at".into(), json!(Utc::now().to_rfc3339()));

    if let Some(desc) = payload["short_description"].as_str() {
        let cleaned = security::sanitize_text(desc, 500);
        body.insert("short_description".into(), json!(cleaned.clone()));
        body.insert("description".into(), json!(cleaned));
    }

    if let Some(items) = payload["known_for"].as_array() {
        let known_for = items
            .iter()
            .filter_map(Value::as_str)
            .filter_map(|value| security::sanitize_optional_text(Some(value), 60))
            .take(12)
            .collect::<Vec<_>>();
        body.insert("known_for".into(), json!(known_for));
    }

    let updated = state
        .db
        .supabase
        .update_json("businesses", &[eq("id", &id)], Value::Object(body))
        .await?;
    let row = updated
        .into_iter()
        .next()
        .ok_or_else(|| AppError::NotFound("Business not found".into()))?;
    Ok(Json(normalize_business(row)))
}

async fn delete_business(
    State(state): State<Arc<AppState>>,
    auth_user: AuthUser,
    Path(id): Path<String>,
) -> Result<impl IntoResponse> {
    let id = security::validate_uuid_id(&id, "business ID")?;
    let existing = find_business(&state, &id).await?;
    ensure_business_owner(&existing, &auth_user)?;

    state
        .db
        .supabase
        .delete_json("businesses", &[eq("id", &id)])
        .await?;

    Ok(Json(json!({ "message": "Business deleted" })))
}

async fn get_business_photo(
    State(state): State<Arc<AppState>>,
    Path(_id): Path<String>,
    Query(params): Query<std::collections::HashMap<String, String>>,
) -> Result<Response> {
    let photo_ref = params
        .get("ref")
        .cloned()
        .ok_or_else(|| AppError::BadRequest("Missing photo ref".into()))?;
    proxy_google_photo(&state, &photo_ref, &params).await
}

async fn get_google_place_photo(
    State(state): State<Arc<AppState>>,
    Query(params): Query<std::collections::HashMap<String, String>>,
) -> Result<Response> {
    let place_id = params
        .get("place_id")
        .map(|value| security::sanitize_text(value, 180))
        .filter(|value| !value.is_empty())
        .ok_or_else(|| AppError::BadRequest("Missing place_id".into()))?;

    if state.config.google_api_key.is_empty() {
        return Err(AppError::BadRequest("Google API not configured".into()));
    }

    let details =
        crate::services::google_places::get_place_details(&state.config.google_api_key, &place_id)
            .await
            .map_err(|e| AppError::Internal(e.to_string()))?;
    let photo_ref = details["photos"]
        .as_array()
        .and_then(|photos| photos.first())
        .and_then(|photo| photo["photo_reference"].as_str())
        .ok_or_else(|| AppError::NotFound("Photo not found".into()))?;

    proxy_google_photo(&state, photo_ref, &params).await
}

async fn proxy_google_photo(
    state: &AppState,
    photo_ref: &str,
    params: &std::collections::HashMap<String, String>,
) -> Result<Response> {
    let photo_ref = security::validate_photo_reference(photo_ref)?;

    if state.config.google_api_key.is_empty() {
        return Err(AppError::BadRequest("Google API not configured".into()));
    }

    let max_width = params
        .get("maxwidth")
        .and_then(|value| value.parse::<u32>().ok())
        .unwrap_or(800)
        .clamp(64, 1600);

    let url = crate::services::google_places::get_photo_url(
        &state.config.google_api_key,
        &photo_ref,
        max_width,
    )
    .await;

    let response = reqwest::get(&url)
        .await
        .map_err(|e| AppError::Internal(e.to_string()))?;
    if !response.status().is_success() {
        return Err(AppError::BadRequest("Photo is unavailable".into()));
    }

    let content_type = response
        .headers()
        .get(header::CONTENT_TYPE)
        .cloned()
        .unwrap_or_else(|| HeaderValue::from_static("image/jpeg"));
    let body: Bytes = response
        .bytes()
        .await
        .map_err(|e| AppError::Internal(e.to_string()))?;

    Ok((
        [
            (header::CONTENT_TYPE, content_type),
            (
                header::CACHE_CONTROL,
                HeaderValue::from_static("public, max-age=86400"),
            ),
        ],
        body,
    )
        .into_response())
}

async fn query_businesses(
    state: &AppState,
    params: &BusinessSearchQuery,
    limit_value: i64,
) -> Result<Vec<Value>> {
    if let (Some(lat), Some(lng)) = (params.lat, params.lng) {
        security::validate_lat_lng(lat, lng)?;
    }

    let mut query: QueryParams = vec![
        select_all(),
        limit(if params.lat.is_some() && params.lng.is_some() {
            limit_value * 3
        } else {
            limit_value
        }),
        offset(params.offset.unwrap_or(0)),
    ];

    if let Some(category) = params.category.as_ref().filter(|value| !value.is_empty()) {
        query.push(eq("category", security::sanitize_text(category, 80)));
    }
    if let Some(owner_id) = params.owner_id.as_ref().filter(|value| !value.is_empty()) {
        query.push(eq(
            "owner_id",
            security::validate_uuid_id(owner_id, "owner ID")?,
        ));
    }
    if params.verified_only.unwrap_or(false) {
        query.push(is_true("is_verified"));
    }
    if let Some(search) = params
        .q
        .as_deref()
        .or(params.search.as_deref())
        .filter(|value| !value.trim().is_empty())
    {
        if let Some(filter) = ilike_or_filter(&["name", "description", "category"], search) {
            query.push(filter);
        }
    }

    let sort = params.sort.as_deref().or(params.sort_by.as_deref());
    match sort {
        Some("rating") => query.push(order("rating.desc.nullslast,review_count.desc")),
        Some("newest") => query.push(order("created_at.desc")),
        Some("name") => query.push(order("name.asc")),
        _ => query.push(order("created_at.desc")),
    }

    let rows = state.db.supabase.select_json("businesses", &query).await?;
    let mut businesses = rows
        .into_iter()
        .map(normalize_business)
        .filter_map(|mut row| {
            if let (Some(user_lat), Some(user_lng)) = (params.lat, params.lng) {
                let (lat, lng) = business_lat_lng(&row)?;
                let radius = params.radius_km.or(params.radius).unwrap_or(25.0);
                let distance =
                    crate::services::geo_service::haversine_km(user_lat, user_lng, lat, lng);
                if distance > radius {
                    return None;
                }
                row["distance"] = json!(distance);
                row["distance_km"] = json!(distance);
            }
            Some(row)
        })
        .collect::<Vec<_>>();

    match sort {
        Some("distance") => businesses.sort_by(|a, b| {
            a["distance_km"]
                .as_f64()
                .unwrap_or(f64::MAX)
                .partial_cmp(&b["distance_km"].as_f64().unwrap_or(f64::MAX))
                .unwrap_or(std::cmp::Ordering::Equal)
        }),
        Some("rating") => businesses.sort_by(|a, b| {
            b["rating"]
                .as_f64()
                .unwrap_or(0.0)
                .partial_cmp(&a["rating"].as_f64().unwrap_or(0.0))
                .unwrap_or(std::cmp::Ordering::Equal)
        }),
        _ => {}
    }

    businesses.truncate(limit_value as usize);
    Ok(businesses)
}

async fn find_business(state: &AppState, id: &str) -> Result<Value> {
    let id = security::validate_uuid_id(id, "business ID")?;
    state
        .db
        .supabase
        .select_one_json("businesses", &[select_all(), eq("id", id)])
        .await?
        .ok_or_else(|| AppError::NotFound("Business not found".into()))
}

fn ensure_business_owner(row: &Value, auth_user: &AuthUser) -> Result<()> {
    let owner_id = value_str(row, "owner_id");
    if owner_id != auth_user.id && auth_user.role != "admin" {
        return Err(AppError::Forbidden("Not the business owner".into()));
    }
    Ok(())
}

fn insert_optional_text(body: &mut Map<String, Value>, key: &str, value: Option<&str>, max: usize) {
    if let Some(value) = value.and_then(|raw| security::sanitize_optional_text(Some(raw), max)) {
        body.insert(key.into(), json!(value));
    }
}
